### PR TITLE
Implement behaviour of yield* as specced by rev32 of the ES6 draft.

### LINKED ISSUE #1688: Implement new yield *-semantics
In Rev 32 the semantics of yield * has been changed. See p. 301 of:

http://wiki.ecmascript.org/lib/exe/fetch.php?id=harmony%3Aspecification_drafts&cache=cache&media=harmony:working_draft_ecma-262_edition_6_02-02-15-rev32markup.pdf

I'll check whether I can do something about it.

(Part of `generator.js` was already ahead of time. Namely parts have already been working as described by rev 32 but was wrong wrt rev 31.)

### PR DESCRIPTION
**Pull request by mnieper**
This should close #1688.

### CODE REVIEW

**Comment by arv** [comment]
You should probably rebase with upstream/master. Your PR has a lot of commits that have been committed already.

File: src/runtime/generators.js [hunk 1]
```diff
@@ -130,14 +131,19 @@ GeneratorContext.prototype = {
...
+            if (!result.done) {
+              ctx.returnValue = ctx.oldReturnValue;
+              return result;
+            }
             ctx.returnValue = result.value;
           }
           throw e;
-        }
-        if (iterator.throw) {
+        } else if (iterator.throw) {
```
Comment (arv) [inline src/runtime/generators.js]: no else after return/throw

File: test/feature/Yield/BadIterable.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        return {value: undefined, done: true};
+      }
+    };
+  }
+}
+
+var i1 = new BadIterable();
+var g1 = wrap(i1);
+assert.deepEqual(g1.next(), {value: 42, done: false});
+assert.throws(() => g1.throw("ex1"), TypeError);
```
Comment (arv) [inline test/feature/Yield/BadIterable.js]: we use single quotes for consistency

File: test/feature/Yield/BadIterable.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+assert.equal(i1.closed, true);
```
Comment (arv) [inline test/feature/Yield/BadIterable.js]: you can do isTrue here

**Comment by arv** [comment]
LGTM

Could you squash these into one commit?

**Comment by mnieper** [comment]
Did the corrections. Have to read about squashing and rebasing. Don't want to wreck my repository as I have a number of branches open.

**Comment by arv** [comment]
If you don't want to rebase/squash I can do that before merging to master.

**Comment by mnieper** [comment]
Sorry, too late here already. Thus the commits are not yet squashed (my first attempt failed) but the upstream master branch has been merged into.

Do you happen to know a quick guide for the squashing and rebasing?

**Comment by johnjbarton** [comment]
When we say "rebase" we mean the equivalent of:
```
git checkout master
git pull --ff-only upstream master  # origin master syncs to upstream.
git checkout <branch>
git rebase master  # branch syncs to master
```
This affects just two branches, master and `<branch>`.  

When we say squash, its like this:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

See also http://eagain.net/articles/git-for-computer-scientists/

HTH

**Comment by UltCombo** [comment]
You can skip pulling the master by rebasing directly on upstream/master btw:

```
# in case you haven't added the upstream remote yet:
git remote add upstream https://github.com/google/traceur-compiler.git
git fetch --all
# on your "yield-star" branch
git rebase upstream/master
```

**Comment by UltCombo** [comment]
By the way, if you're going to try to squash commits for the first time, create another branch from your current branch first:

```
git checkout -b test
```

Because odds that something will go terribly wrong are quite high (talking from own experience), so if the worst happens you can just `git checkout yield-star` and create another branch to try again.

**Comment by mnieper** [comment]
Rebased and squashed, although the result does not look perfect. :-(

@arv So please do any further squashing that needs to be done; should be ready for merge.

**Comment by arv** [comment]
Committed as 28baf5195da2905b5ff3fc2a6456dd0c86e7e514

I ended up reconstructing the commit.

```sh
git reset master
git add ...
git commit --author="..."
```

**Comment by mnieper** [comment]
Thanks. Closing this.

**Pull request closed without merging**